### PR TITLE
Added Feature variables and its testcases

### PR DESCRIPTION
**Pull request by mnoman09**

### CODE REVIEW

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 import com.optimizely.ab.bucketing.Bucketer;
 import com.optimizely.ab.bucketing.DecisionService;
 import com.optimizely.ab.config.Experiment;
+import com.optimizely.ab.config.LiveVariable;
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: Remove unused imports

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 2]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+
+    //=======Feature Variables Testing===========
+    @Test
+    public void testGoodGetFeatureVariableBoolean() {
+        String featureKey = "boolean_single_variable_feature";
+        String variableKey = "boolean_variable";
+        OptimizelyClient optimizelyClient = new OptimizelyClient(optimizelyWithV4Datafile, logger);
+
+        //Scenario#1 Without attributes
+        assertTrue(optimizelyClient.getFeatureVariableBoolean(featureKey, variableKey,genericUserId));
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: Is there a way to check if getFeatureVariableBoolean returns false? If variable value is false then this case will fail.
Is there a way to make sure it's true?

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 3]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+
+    @Test
+    public void testBadGetFeatureVariableBoolean() {
+        String featureKey = "boolean_single_variable_feature";
+        String variableKey = "boolean_variable";
+
+        OptimizelyClient optimizelyClient = new OptimizelyClient(null, logger);
+
+        //Scenario#1 without attributes
+        optimizelyClient.getFeatureVariableBoolean(featureKey,variableKey,genericUserId);
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: assertNull here

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 4]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+    public void testGoodGetFeatureVariableBoolean() {
+        String featureKey = "boolean_single_variable_feature";
+        String variableKey = "boolean_variable";
+        OptimizelyClient optimizelyClient = new OptimizelyClient(optimizelyWithV4Datafile, logger);
+
+        //Scenario#1 Without attributes
+        assertTrue(optimizelyClient.getFeatureVariableBoolean(featureKey, variableKey,genericUserId));
+
+        //Scenario#2 With attributes
+        assertTrue(optimizelyClient.getFeatureVariableBoolean(featureKey,
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: Is there a way to check if getFeatureVariableBoolean returns false? If variable value is false then this case will fail.
Is there a way to make sure it's true?

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 5]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+        String variableKey = "boolean_variable";
+
+        OptimizelyClient optimizelyClient = new OptimizelyClient(null, logger);
+
+        //Scenario#1 without attributes
+        optimizelyClient.getFeatureVariableBoolean(featureKey,variableKey,genericUserId);
+        verify(logger).warn("Optimizely is not initialized, could not get feature {} variable {} boolean for user {}", featureKey                                   ,variableKey, genericUserId);
+
+        //Scenario#2 with attributes
+        optimizelyClient.getFeatureVariableBoolean(featureKey,variableKey,genericUserId,Collections.<String, String>emptyMap());
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: assertNull here

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 6]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+        String variableKey = "double_variable";
+        OptimizelyClient optimizelyClient = new OptimizelyClient(optimizelyWithV4Datafile, logger);
+
+        //Scenario#1 without attributes
+        assertEquals(14.99,optimizelyClient.getFeatureVariableDouble(
+                featureKey,
+                variableKey,
+                genericUserId
+        ));
+
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: verifyZeroInteractions(logger); here too

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 7]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+    @Test
+    public void testBadGetFeatureVariableDouble() {
+        String featureKey = "double_single_variable_feature";
+        String variableKey = "double_variablex";
+
+        OptimizelyClient optimizelyClient = new OptimizelyClient(null, logger);
+
+        //Scenario#1 without attributes
+        optimizelyClient.getFeatureVariableDouble(featureKey,variableKey,genericUserId);
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: assertNull here

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 8]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+        verify(logger).warn("Optimizely is not initialized, could not get feature {} variable {} double for user {}", featureKey                                    ,variableKey, genericUserId);
+
+        //Scenario#2 with attributes
+        optimizelyClient.getFeatureVariableDouble(featureKey,variableKey,genericUserId,Collections.<String, String>emptyMap());
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: assertNull here

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 9]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+    }
+
+    @Test
+    public void testGoodGetFeatureVariableInteger() {
+        String featureKey = "integer_single_variable_feature";
+        String variableKey = "integer_variable";
+        OptimizelyClient optimizelyClient = new OptimizelyClient(optimizelyWithV4Datafile, logger);
+
+        //Scenario#1 without attributes
+        assertEquals((long) 7,(long) optimizelyClient.getFeatureVariableInteger(
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: We can do this here assertEquals( 7,(int) optimizelyClient.getFeatureVariableInteger(

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java [hunk 10]
```diff
@@ -849,4 +855,155 @@ public void testBadClearNotificationListeners() {
...
+                featureKey,
+                variableKey,
+                genericUserId
+        ));
+
+        //Scenario#2 with attributes
+        assertEquals((long)7, (long)optimizelyClient.getFeatureVariableInteger(
```
Comment (raheeladnan) [inline android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: We can do this here assertEquals( 7,(int) optimizelyClient.getFeatureVariableInteger(

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java (hunk 8 above)
Comment (raheeladnan) [reply android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: Do this for all the methods which are to test Bad feature variables below.

File: android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java (hunk 2 above)
Comment (mnoman09) [reply android-sdk/src/androidTest/java/com/optimizely/ab/android/sdk/OptimizelyClientTest.java]: no its not possible for it to return false with given datafile so it wont fail

**Pull request merged**